File: debop4s-timeperiod/src/test/scala/debop4s/timeperiod/tests/timeranges/HourRangeTest.scala
package debop4s.timeperiod.tests.timeranges

import debop4s.core.jodatime._
import debop4s.timeperiod._
import debop4s.timeperiod.tests.AbstractTimePeriodTest
import debop4s.timeperiod.timerange.HourRange
import debop4s.timeperiod.utils.Times


/**
 * HourRangeTest
 * Created by debop on 2014. 2. 16.
 */
class HourRangeTest extends AbstractTimePeriodTest {

    test("init values") {
        val now = Times.now
        val firstHour = Times.trimToMinute(now)
        val secondHour = firstHour + 1.hour

        val hourRange = HourRange(now, EmptyOffsetTimeCalendar)

        hourRange.start.getYear should equal(firstHour.getYear)
        hourRange.start.getMonthOfYear should equal(firstHour.getMonthOfYear)
        hourRange.start.getDayOfMonth should equal(firstHour.getDayOfMonth)
        hourRange.start.getHourOfDay should equal(firstHour.getHourOfDay)
        hourRange.start.getMinuteOfHour should equal(0)
        hourRange.start.getSecondOfMinute should equal(0)
        hourRange.start.getMillisOfSecond should equal(0)

        hourRange.end.getYear should equal(secondHour.getYear)
        hourRange.end.getMonthOfYear should equal(secondHour.getMonthOfYear)
        hourRange.end.getDayOfMonth should equal(secondHour.getDayOfMonth)
        hourRange.end.getHourOfDay should equal(secondHour.getHourOfDay)
        hourRange.end.getMinuteOfHour should equal(0)
        hourRange.end.getSecondOfMinute should equal(0)
        hourRange.end.getMillisOfSecond should equal(0)
    }

    test("default calendar") {
        val now = Times.now
        val today = Times.today

        for (h <- 0 until HoursPerDay) {
            val hr = HourRange(today + h.hour)

            hr.year should equal(today.getYear)
            hr.monthOfYear should equal(today.getMonthOfYear)
            hr.dayOfMonth should equal(today.getDayOfMonth)
            hr.hourOfDay should equal(h)
            hr.start should equal(hr.calendar.mapStart(today + h.hour))
            hr.end should equal(hr.calendar.mapEnd(today + (h + 1).hour))
        }
    }

    test("HourRange constructor test") {
        val now = Times.now

        val hr = HourRange(now)
        hr.year should equal(now.getYear)
        hr.monthOfYear should equal(now.getMonthOfYear)
        hr.dayOfMonth should equal(now.getDayOfMonth)
        hr.hourOfDay should equal(now.getHourOfDay)

        val hr2 = HourRange(now.getYear, now.getMonthOfYear, now.getDayOfMonth, now.getHourOfDay)

        hr2.year should equal(now.getYear)
        hr2.monthOfYear should equal(now.getMonthOfYear)
        hr2.dayOfMonth should equal(now.getDayOfMonth)
        hr2.hourOfDay should equal(now.getHourOfDay)
    }

    test("add hour") {
        val hr = HourRange()
        hr.previousHour.hourOfDay should equal(hr.start.plusHours(-1).getHourOfDay)
        hr.nextHour.hourOfDay should equal(hr.start.plusHours(1).getHourOfDay)

        val hr2 = HourRange(EmptyOffsetTimeCalendar)
        hr2.addHours(0) should equal(hr2)

        val prevRange = hr2.previousHour
        prevRange.year should equal(hr2.addHours(-1).year)
        prevRange.monthOfYear should equal(hr2.addHours(-1).monthOfYear)
        prevRange.dayOfMonth should equal(hr2.addHours(-1).dayOfMonth)
        prevRange.hourOfDay should equal(hr2.addHours(-1).hourOfDay)

        val nextRange = hr2.nextHour
        nextRange.year should equal(hr2.addHours(1).year)
        nextRange.monthOfYear should equal(hr2.addHours(1).monthOfYear)
        nextRange.dayOfMonth should equal(hr2.addHours(1).dayOfMonth)
        nextRange.hourOfDay should equal(hr2.addHours(1).hourOfDay)

        for (h <- -100 to 100 by 5) {
            val r = hr2.addHours(h)
            val r2 = hr2.addHours(h)

            r.year should equal(r2.year)
            r.monthOfYear should equal(r2.monthOfYear)
            r.dayOfMonth should equal(r2.dayOfMonth)
            r.hourOfDay should equal(r2.hourOfDay)
        }
    }

    test("getMinutes") {
        val hr = HourRange()
        val minutes = hr.minutes

        minutes.size should equal(MinutesPerHour)

        for (i <- 0 until MinutesPerHour) {
            val minute = minutes(i)
            minute.start should equal(hr.start + i.minute)
            minute.unmappedStart should equal(hr.start + i.minute)

            minute.end should equal(minute.calendar.mapEnd(hr.start + (i + 1).minute))
            minute.unmappedEnd should equal(hr.start + (i + 1).minute)
        }
    }


}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/HourRangeTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.HourRange;
import debop4s.timeperiod.timerange.MinuteRange;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

@Slf4j
public class HourRangeTest extends TimePeriodTestBase {

    @Test
    public void initValues() {
        DateTime now = Times.now();
        DateTime firstHour = Times.trimToMinute(now);
        DateTime secondHour = firstHour.plusHours(1);

        HourRange hourRange = new HourRange(now, TimeCalendar.getEmptyOffset());

        assertThat(hourRange.getStart().getYear()).isEqualTo(firstHour.getYear());
        assertThat(hourRange.getStart().getMonthOfYear()).isEqualTo(firstHour.getMonthOfYear());
        assertThat(hourRange.getStart().getDayOfMonth()).isEqualTo(firstHour.getDayOfMonth());
        assertThat(hourRange.getStart().getMinuteOfHour()).isEqualTo(0);
        assertThat(hourRange.getStart().getSecondOfMinute()).isEqualTo(0);
        assertThat(hourRange.getStart().getMillisOfSecond()).isEqualTo(0);

        assertThat(hourRange.getEnd().getYear()).isEqualTo(secondHour.getYear());
        assertThat(hourRange.getEnd().getMonthOfYear()).isEqualTo(secondHour.getMonthOfYear());
        assertThat(hourRange.getEnd().getDayOfMonth()).isEqualTo(secondHour.getDayOfMonth());
        assertThat(hourRange.getEnd().getMinuteOfHour()).isEqualTo(0);
        assertThat(hourRange.getEnd().getSecondOfMinute()).isEqualTo(0);
        assertThat(hourRange.getEnd().getMillisOfSecond()).isEqualTo(0);
    }

    @Test
    public void defaultCalendar() {
        DateTime now = Times.now();
        DateTime today = Times.today();

        for (int h = 0; h < TimeSpec.HoursPerDay; h++) {
            HourRange hourRange = new HourRange(today.plusHours(h));

            assertThat(hourRange.getYear()).isEqualTo(today.getYear());
            assertThat(hourRange.getMonthOfYear()).isEqualTo(today.getMonthOfYear());
            assertThat(hourRange.getDayOfMonth()).isEqualTo(today.getDayOfMonth());
            assertThat(hourRange.getHourOfDay()).isEqualTo(h);
            assertThat(hourRange.getStart()).isEqualTo(hourRange.getTimeCalendar().mapStart(today.plusHours(h)));
            assertThat(hourRange.getEnd()).isEqualTo(hourRange.getTimeCalendar().mapEnd(today.plusHours(h + 1)));
        }
    }

    @Test
    public void constructorTest() {
        DateTime now = Times.now();

        HourRange hourRange = new HourRange(now);
        assertThat(hourRange.getYear()).isEqualTo(now.getYear());
        assertThat(hourRange.getMonthOfYear()).isEqualTo(now.getMonthOfYear());
        assertThat(hourRange.getDayOfMonth()).isEqualTo(now.getDayOfMonth());
        assertThat(hourRange.getHourOfDay()).isEqualTo(now.getHourOfDay());

        hourRange = new HourRange(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), now.getHourOfDay());
        assertThat(hourRange.getYear()).isEqualTo(now.getYear());
        assertThat(hourRange.getMonthOfYear()).isEqualTo(now.getMonthOfYear());
        assertThat(hourRange.getDayOfMonth()).isEqualTo(now.getDayOfMonth());
        assertThat(hourRange.getHourOfDay()).isEqualTo(now.getHourOfDay());
    }

    @Test
    public void addHourTest() {
        HourRange hourRange = new HourRange();

        assertThat(hourRange.previousHour().getHourOfDay()).isEqualTo(hourRange.getStart().plusHours(-1).getHourOfDay());
        assertThat(hourRange.nextHour().getHourOfDay()).isEqualTo(hourRange.getStart().plusHours(1).getHourOfDay());

        hourRange = new HourRange(TimeCalendar.getEmptyOffset());

        assertThat(hourRange.addHours(0)).isEqualTo(hourRange);

        HourRange prevRange = hourRange.previousHour();
        assertThat(hourRange.addHours(-1).getYear()).isEqualTo(prevRange.getYear());
        assertThat(hourRange.addHours(-1).getMonthOfYear()).isEqualTo(prevRange.getMonthOfYear());
        assertThat(hourRange.addHours(-1).getDayOfMonth()).isEqualTo(prevRange.getDayOfMonth());
        assertThat(hourRange.addHours(-1).getHourOfDay()).isEqualTo(prevRange.getHourOfDay());

        HourRange nextRange = hourRange.nextHour();
        assertThat(hourRange.addHours(1).getYear()).isEqualTo(nextRange.getYear());
        assertThat(hourRange.addHours(1).getMonthOfYear()).isEqualTo(nextRange.getMonthOfYear());
        assertThat(hourRange.addHours(1).getDayOfMonth()).isEqualTo(nextRange.getDayOfMonth());
        assertThat(hourRange.addHours(1).getHourOfDay()).isEqualTo(nextRange.getHourOfDay());

        for (int h = -100; h < 100; h += 5) {
            HourRange range = hourRange.addHours(h);
            HourRange range2 = hourRange.addHours(h);
            assertThat(range.getYear()).isEqualTo(range2.getYear());
            assertThat(range.getMonthOfYear()).isEqualTo(range2.getMonthOfYear());
            assertThat(range.getDayOfMonth()).isEqualTo(range2.getDayOfMonth());
            assertThat(range.getHourOfDay()).isEqualTo(range2.getHourOfDay());
        }
    }

    @Test
    public void getMinutesTest() {
        HourRange hourRange = new HourRange();
        List<MinuteRange> minutes = hourRange.getMinutes();

        assertThat(minutes.size()).isEqualTo(TimeSpec.MinutesPerHour);

        for (int i = 0; i < TimeSpec.MinutesPerHour; i++) {
            MinuteRange minute = minutes.get(i);

            assertThat(minute.getStart()).isEqualTo(hourRange.getStart().plusMinutes(i));
            assertThat(minute.getUnmappedStart()).isEqualTo(hourRange.getStart().plusMinutes(i));

            assertThat(minute.getEnd()).isEqualTo(minute.getTimeCalendar().mapEnd(hourRange.getStart().plusMinutes(i + 1)));
            assertThat(minute.getUnmappedEnd()).isEqualTo(hourRange.getStart().plusMinutes(i + 1));
        }
    }
}
